src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -3048,6 +3044,63 @@ ErrorIfCannotPushdownSubquery(Query *subqueryTree, bool outerQueryHasLimit)
 
 
 /*
+ * ErrorIfUnsupportedUnionQuery is a helper function for ErrorIfCannotPushdownSubquery().
+ * It basically iterates over the subqueries that reside under the given set operations.
+ *
+ * The function also errors out for set operations INTERSECT and EXCEPT.
+ */
+static void
+ErrorIfUnsupportedUnionQuery(List *rangeTableList,
+							 SetOperationStmt *setOperationStatement,
+							 bool outerQueryHasLimit)
+{
+	Node *leftArgument = setOperationStatement->larg;
+	Node *rightArgument = setOperationStatement->rarg;
+
+	if (setOperationStatement->op != SETOP_UNION)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot push down this subquery"),
+						errdetail("Intersect and Except are currently unsupported")));
+	}
+
+	if (IsA(leftArgument, RangeTblRef))
+	{
+		RangeTblRef *rangeTableReference = (RangeTblRef *) leftArgument;
+		RangeTblEntry *rangeTableEntry = rt_fetch(rangeTableReference->rtindex,
+												  rangeTableList);
+
+		Assert(rangeTableEntry->rtekind == RTE_SUBQUERY);
+
+		ErrorIfCannotPushdownSubquery(rangeTableEntry->subquery, outerQueryHasLimit);
+	}
+	else if (IsA(leftArgument, SetOperationStmt))
+	{
+		SetOperationStmt *leftUnion = (SetOperationStmt *) leftArgument;
+
+		ErrorIfUnsupportedUnionQuery(rangeTableList, leftUnion, outerQueryHasLimit);
```

Comment:
Why is so? I'd be in favor of processing in the format that postgres keeps these structs.

> We should also think about removing indirect recursion

Again, Postgres keeps exactly in that format. Why should we not follow it?